File: src/SummonerEmote.ts
export default class SummonerEmote {
  id: number;
  name: string;
  inventoryIcon: string;
  description: string;

  constructor({
    id, name, inventoryIcon, description
  }: IEmoteData) {
    this.id = id;
    this.name = name;
    this.inventoryIcon = inventoryIcon;
    this.description = description;
  }

  static fromJson(data: IEmoteData) {
    if (data == null) throw new Error("data must be defined");

    const emote = new SummonerEmote(data);
    return emote;
  }

  get uri() {
    const { inventoryIcon } = this;
    if (inventoryIcon == null) return "";

    const split = inventoryIcon.split("/lol-game-data/assets").filter(x => x);
    if (split.length == 1) return split[0].toLowerCase();

    return "";
  }

  get categories() {
    if (!this.isIcon) return [];

    const split = this.inventoryIcon.split("/lol-game-data/assets/ASSETS/Loadouts/SummonerEmotes/").filter(x => x);
    const subsplit = split[0].replace("_", "").split("/").filter(x => x);

    const categories = [];
    console.log(subsplit);
    for (const category of subsplit) {
      if (!category.includes(".png")) {
        const regex = /(?<=[A-Z])(?=[A-Z][a-z])|(?<=[^A-Z])(?=[A-Z])|(?<=[A-Za-z])(?=[^A-Za-z])/;
        const formatted = category.split(regex);
        if (formatted != null) categories.push(formatted[0]);
      }
    }
    console.log(categories);

    return categories;
  }

  get filename() {
    const { inventoryIcon } = this;
    if (inventoryIcon == null) return "";

    const filename = `${inventoryIcon.split("/").slice(-1)}`;
    const subnames = filename.split("_");
    if (subnames.slice(0, 1).includes(`${this.id}`)) return filename;
    return `${this.id}_${filename}`;
  }

  get isIcon() {
    return this.uri == null || this.uri == "" || this.uri == "/" ? false : true;
  }
}

interface IEmoteData {
  id: number;
  name: string;
  inventoryIcon: string;
  description: string;
}

File: Models/SummonerEmote.cs
using System.Collections.Immutable;
using System.Text.RegularExpressions;

namespace Models;

public partial class SummonerEmote 
{
    public uint Id { get; set; }
    public string Name { get; set; } = "";
    public string InventoryIcon { get; set; } = "";
    public string Description { get; set; } = "";

    public string Uri 
    { 
        get 
        {
            if (string.IsNullOrEmpty(InventoryIcon)) return "";

            var split = InventoryIcon.Split("/lol-game-data/assets", StringSplitOptions.RemoveEmptyEntries);
            if (split.Length == 1) return split[0].ToLower();
            return "";
        }
    }

    public bool IsIcon 
    {
        get => string.IsNullOrEmpty(Uri) || Uri.Equals("/") ? false : true;
    }

    public string FileName 
    {
        get 
        {
            if (string.IsNullOrEmpty(InventoryIcon)) return "";

            var fileName = $"{InventoryIcon.Split("/").Last()}";
            var subnames = fileName.Split("_");
            if (subnames.First().Contains($"{Id}")) return fileName;
            return $"{Id}_{fileName}";
        }
    }

    public ImmutableList<string> Categories 
    {
        get 
        {
            if (!IsIcon) return ImmutableList<string>.Empty;

            var split = InventoryIcon
                .Split("/lol-game-data/assets/ASSETS/Loadouts/SummonerEmotes/", StringSplitOptions.RemoveEmptyEntries);
            var subsplit = split
                .First()
                .Replace("_", "")
                .Split("/", StringSplitOptions.RemoveEmptyEntries);
            
            var categories = new List<string>();
            foreach (var category in subsplit) 
            {
                if (!category.Contains(".png"))
                {
                    var regex = CategoryFilterRegex();
                    var formatted = regex.Split(category);
                    categories.Add(string.Join(" ", formatted));
                }
            }
            return categories.ToImmutableList();
        }
    }

    [GeneratedRegex("(?<=[A-Z])(?=[A-Z][a-z])|(?<=[^A-Z])(?=[A-Z])|(?<=[A-Za-z])(?=[^A-Za-z])")]
    private static partial Regex CategoryFilterRegex();
}